xarray/core/dataset.py

```diff
@@ -1522,6 +1522,11 @@ def _setitem_check(self, key, value):
                         f"Variable '{name}': dimension order differs between"
                         f" original and new data:\n{dims}\nvs.\n{val.dims}"
                     )
+            else:
+                val = np.array(val)
+
+            # check type consistency
+            val.astype(var_k.dtype)
```

Comment:
Can we use this modified version as the value to set? Casting might be expensive, it would be nice not to need to do it twice.

E.g., `val = val.astype(var_k.dtype, copy=False)`

Reply:
ok, good point. I implemented this.